pyi.py

```diff
@@ -303,6 +304,85 @@ def _is_object(node: ast.expr, name: str, *, from_: Container[str]) -> bool:
 _is_final = partial(_is_object, name="final", from_=_TYPING_MODULES)
 _is_Final = partial(_is_object, name="Final", from_=_TYPING_MODULES)
 _is_Self = partial(_is_object, name="Self", from_=({"_typeshed"} | _TYPING_MODULES))
+_is_TracebackType = partial(_is_object, name="TracebackType", from_={"types"})
+
+
+def _is_type_or_Type(node: ast.expr) -> bool:
+    """
+    >>> import ast
+    >>> ast_node_for = lambda string: ast.parse(string).body[0].value
+    >>> _is_type_or_Type(ast_node_for('type'))
+    True
+    >>> _is_type_or_Type(ast_node_for('Type'))
+    True
+    >>> _is_type_or_Type(ast_node_for('builtins.type'))
+    True
+    >>> _is_type_or_Type(ast_node_for('typing_extensions.Type'))
+    True
+    >>> _is_type_or_Type(ast_node_for('typing.Type'))
+    True
+    """
+    if isinstance(node, ast.Name):
+        return node.id in {"type", "Type"}
+    if isinstance(node, ast.Attribute):
+        node_value = node.value
+        if not isinstance(node_value, ast.Name):
+            return False
+        node_value_id = node_value.id
+        attr = node.attr
+        return (node_value_id == "builtins" and attr == "type") or (
+            node_value_id in _TYPING_MODULES and attr == "Type"
+        )
+    return False
+
+
+def _is_PEP_604_union(node: ast.expr | None) -> TypeGuard[ast.BinOp]:
+    return isinstance(node, ast.BinOp) and isinstance(node.op, ast.BitOr)
+
+
+def _is_None(node: ast.expr) -> bool:
+    # <=3.7: `BaseException | None` parses as BinOp(left=Name(id='BaseException'), op=BitOr(), right=NameConstant(value=None))`
+    # >=3.8: `BaseException | None` parses as BinOp(left=Name(id='BaseException'), op=BitOr(), right=Constant(value=None))`
+    # ast.NameConstant is deprecated in 3.8+, but doesn't raise a DeprecationWarning (and the isinstance() check still works)
+    return isinstance(node, ast.NameConstant) and node.value is None
+
+
+_sentinel = ast.expr()
+
+
+class ExitArgAnalysis(NamedTuple):
+    is_union_with_None: bool
+    non_None_part: ast.expr = _sentinel
+
+    def __repr__(self) -> str:
+        return (
+            f"ExitArgAnalysis("
+            f"is_union_with_None={self.is_union_with_None}, "
+            f"non_None_part={ast.dump(self.non_None_part)}"
+            f")"
+        )
+
+
+def _analyse_exit_method_arg(node: ast.BinOp) -> ExitArgAnalysis:
+    """Return a two-item tuple providing analysis of the annotation of an exit-method arg.
+
+    This function is only called if we already know
+    that `node` is a `BoolOp` and node.op is a `BitOr`.
```

Comment:
I would prefer `assert isinstance(node.op, ast.BitOr)` in the method body to communicate this.